### ISSUE TITLE
dbus error with large number of filesystems in a pool

### ISSUE THREAD

**Issue by trgill** [stratis-storage/stratisd#1426]
The attached script fails with the following error:

```
+ stratis filesystem snapshot pool1 341 341-snapshot
Execution failure caused by:
Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
    which in turn caused:
Error while invoking method "SnapshotFilesystem" belonging to interface "org.storage.stratis1.pool" with arguments (/org/storage/stratis1/344, 341-snapshot)
```

```
#!/bin/bash
set -x
set -e
DEVS1="/dev/xxx /dev/xxx"

FS_COUNT=500
POOL_NAME=pool1

~/zero_disks.sh

stratis pool create $POOL_NAME $DEVS1

for((fs=1;fs<=$FS_COUNT;fs++))
do
   echo "create $POOL_NAME-fs-$fs"
   stratis filesystem create $POOL_NAME "$fs"
done

for((fs=1;fs<=$FS_COUNT;fs++))
do
   # remove stale directory if there is one
   rm -rf /tmp/mp-$fs
   # create mount point
   mkdir /tmp/mp-$fs
   stratis filesystem snapshot $POOL_NAME $fs $fs-snapshot
   # mount the source filesystem and run IO
   mount /stratis/$POOL_NAME/$fs /tmp/mp-$fs
done

for((fs=1;fs<=$FS_COUNT;fs++))
do
   fio --filename=/tmp/mp-$fs/fio-test-file ~/basic-write.fio &
done

echo "wait for the fio processes to finish"
wait

echo "start cleanup"
for((fs=1;fs<=$FS_COUNT;fs++))
do
   umount /tmp/mp-$fs
   rm -rf /tmp/mp-$fs
done

for((fs=1;fs<=$FS_COUNT;fs++))
do
   echo "destroy $POOL_NAME-fs-$fs"
   stratis filesystem destroy $POOL_NAME "$fs"
   stratis filesystem destroy $POOL_NAME "$fs-snapshot"
done

stratis pool destroy pool1
```

contents of basic-verify.fio:

```
# cat basic-verify.fio 
[write-and-verify]
rw=randwrite
bs=4k
buffered=0
ioengine=posixaio
fallocate=none
iodepth=16
verify=crc32c
size=2G
max_latency=15s

```

**Comment by mulkieran** [comment]
Can you tell whether the timeout simply expired? Our timeout is libdbus default which is approx 25 seconds. Alternatively, engine failed to return at all.

**Comment by tasleson** [comment]
@mulkieran @trgill  The command line timeout should be 120 seconds:
https://github.com/stratis-storage/stratis-cli/commit/82929a0ec475f7d15b5cc30d120c2c1b200d65cc

**Comment by mulkieran** [comment]
> @mulkieran @trgill The command line timeout should be 120 seconds:
> [stratis-storage/stratis-cli@82929a0](https://github.com/stratis-storage/stratis-cli/commit/82929a0ec475f7d15b5cc30d120c2c1b200d65cc)

Aaah, yes, now I remember that change.

**Comment by mulkieran** [comment]
We do not know whether the timeout occurred in the first stage of snapshotting, i.e., invoking ```GetManagedObjects```, or the second, i.e., when actually invoking the ```SnapshotFilesystem``` method. Either is possible.

**Comment by mulkieran** [comment]
@trgill Thanks for the info. We have confirmed that it is in the ```GetManagedObjects``` call.

**Comment by mulkieran** [comment]
@trgill Can you rerun the test w/ two PRs attached. https://github.com/stratis-storage/stratis-cli/pull/231 just adds primitive timing instrumentation to the CLI. https://github.com/stratis-storage/stratisd/pull/1211 caches the filesystem and pool values that currently require external calls.

With those two PRs in place, we can confirm that it is the external calls that are the culprit.

**Comment by trgill** [comment]
The time for the calls didn't vary much - the call before the one that failed was in the same ballpark as the previous calls.

```
+ stratis --propagate filesystem snapshot pool1 271 271-snapshot
Elapsed time for GetManagedObjects call = 0.130477
Elapsed time for SnapshotFilesystem call = 1.660414
+ mount /stratis/pool1/271 /tmp/mp-271
+ (( fs++ ))
+ (( fs<=500 ))
+ rm -rf /tmp/mp-272
+ mkdir /tmp/mp-272
+ stratis --propagate filesystem snapshot pool1 272 272-snapshot
Traceback (most recent call last):
  File "/root/workspace/dbus-python-client-gen/src/dbus_python_client_gen/_invokers.py", line 319, in dbus_func
    return dbus_method(*xformed_args, timeout=timeout)
  File "/usr/lib64/python3.6/site-packages/dbus/proxies.py", line 145, in __call__
    **keywords)
  File "/usr/lib64/python3.6/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/stratis_cli-1.0.0-py3.6.egg/stratis_cli/_main.py", line 48, in the_func
  File "/usr/local/lib/python3.6/site-packages/stratis_cli-1.0.0-py3.6.egg/stratis_cli/_actions/_util.py", line 105, in the_func
  File "/usr/local/lib/python3.6/site-packages/stratis_cli-1.0.0-py3.6.egg/stratis_cli/_actions/_logical.py", line 125, in snapshot_filesystem
  File "/root/workspace/dbus-python-client-gen/src/dbus_python_client_gen/_invokers.py", line 327, in dbus_func
    name, xformed_args)) from err
dbus_python_client_gen._errors.DPClientInvocationError: Error while invoking method "GetManagedObjects" belonging to interface "org.freedesktop.DBus.ObjectManager" with arguments ()

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/stratis", line 4, in <module>
    __import__('pkg_resources').run_script('stratis-cli==1.0.0', 'stratis')
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 654, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1441, in run_script
    exec(script_code, namespace, namespace)
  File "/usr/local/lib/python3.6/site-packages/stratis_cli-1.0.0-py3.6.egg/EGG-INFO/scripts/stratis", line 33, in <module>
  File "/usr/local/lib/python3.6/site-packages/stratis_cli-1.0.0-py3.6.egg/EGG-INFO/scripts/stratis", line 29, in main
  File "/usr/local/lib/python3.6/site-packages/stratis_cli-1.0.0-py3.6.egg/stratis_cli/_main.py", line 52, in the_func
stratis_cli._errors.StratisCliActionError: Action selected by command-line arguments ['--propagate', 'filesystem', 'snapshot', 'pool1', '272', '272-snapshot'] which were parsed to Namespace(func=<function LogicalActions.snapshot_filesystem at 0x7f2d7f8d0bf8>, origin_name='272', pool_name='pool1', propagate=True, snapshot_name='272-snapshot') failed

```

**Comment by mulkieran** [comment]
Things we are pretty sure of:
* It seems to always fail w/ ```GetManagedObjects```.
* The time for executing ```GetManageObjects``` is about 1/10th that for executing SnapshotFilesystem, indicating that ```SnapshotFilesystem```s calls to xfs are probably taking up the bulk of the time.
* ```GetManagedObjects``` call time does not vary significantly w/ increasing number of filesystems.
* If you retry the snapshot call after a reasonable interval it completes without trouble.

We could attribute this problem to D-Bus itself or to some failure of stratisd, but we can leave xfs out of it.

This test was run w/ both the additions to stratisd to remove ioctl's from D-Bus property methods, and the additions to stratis-cli to record time taken by D-Bus calls.

**Comment by trgill** [comment]
I ran the script again last night.  It completed creating 500 filesystems/500 snapshots - but when I tried to list the pool: 
```

 # stratis pool list
Execution failure caused by:
Message recipient disconnected from message bus without replying
    which in turn caused:
Error while invoking method "GetManagedObjects" belonging to interface "org.freedesktop.DBus.ObjectManager" with arguments ()

```
stratisd output when I ran the command:

```
thread 'main' panicked at 'Kernel must return at least 8 values from thin pool status', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/devicemapper-0.26.1/src/thinpooldev.rs:488:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by mulkieran** [comment]
@trgill: Can you update the version of devicemapper to the most current version from HEAD? That panic was replaced by a more informative error message in a PR that got merged just a few days ago: https://github.com/stratis-storage/devicemapper-rs/pull/417.

**Comment by tasleson** [comment]
Once we have these 1000 dm block devices we consistently get this error, yes?  Trying to understand if the output from device mapper IOCTL always parses incorrectly or if it's a transient error.

**Comment by trgill** [comment]
@tasleson it is a transient error. 

@mulkieran ok will update the dm-rs.  Yes - this test was running off master with a debug build.

**Comment by tasleson** [comment]
@trgill @mulkieran I think we should modify devicemapper-rs to dump the entire results returned from the IOCTL call in this case, so that we can open a bug against dm.

**Comment by trgill** [comment]
Note: running the following commands a few times while the script is running seems to make the problem more likely:

```
stratis pool list
stratis filesystem list
```

**Comment by mulkieran** [comment]
@bgurney-rh Please consider whether you want to try to reproduce or close. stratisd would not be expected to panic anymore, since we released a fix for that panic a few versions back.

**Comment by bgurney-rh** [comment]
I'm going to try a test run of it now, from a September 16 head build, chained off of this stratisd commit:
```
bb1c062 2019-09-16 15:30:40 -0400 :: Merge pull request #1627 from mulkieran/master-changelog
```

**Comment by mulkieran** [comment]
Note: in standup we discussed raising the D-Bus timeout to a very large number, so that the CLI can keep running and exercising the daemon.

**Comment by bgurney-rh** [comment]
I ran the test above on a recent version of stratis, and while it does cause a point where the load average of the system goes up to about 450, and oom-killer terminates about 150 of the fio jobs, it completes successfully.

Note that the original test failure was here:
```
+ stratis filesystem snapshot pool1 341 341-snapshot
```
...and also another exception when trying to run "stratis pool list".  I don't see any problems running the snapshots, nor with running "stratis pool list" during the test, when it has about 500 to 1000 devices, prior to the write.

**Comment by mulkieran** [comment]
@bgurney-rh What do you think of just closing this one?

**Comment by bgurney-rh** [comment]
I'm okay with closing it for now; I was waiting for consensus, but it seems that we have it.

I'll keep this test around, but the "failure" condition seemed more a factor of stalling out the system.  On an attempt with 50 filesystems in the test pool, I saw the test system use 20 GB of cache used in "vmstat 1" output; 100 filesystems in the test pool resulted in 40 GB of cache used.  The test system had a total of 128 GB of memory, and with 500 filesystems, it probably would have used 200 GB of cache, if it could.

But even in those conditions, I did not see the errors reported earlier, so I feel good about closing this issue.
